src/handlers/egress-tracker.js

```diff
@@ -0,0 +1,117 @@
+import { Accounting } from '../services/accounting.js'
+
+/**
+ * @import { Context, IpfsUrlContext, Middleware } from '@web3-storage/gateway-lib'
+ * @import { Environment } from './egress-tracker.types.js'
+ * @import { AccountingService } from '../bindings.js'
+ * @typedef {IpfsUrlContext & { ACCOUNTING_SERVICE?: AccountingService }} EgressTrackerContext
+ */
+
+/**
+ * The egress tracking handler must be enabled after the rate limiting handler,
+ * and before any handler that serves the response body. It uses the CID of the
+ * served content to record the egress in the accounting service, and it counts
+ * the bytes served with a TransformStream to determine the egress amount.
+ *
+ * @type {Middleware<EgressTrackerContext, EgressTrackerContext, Environment>}
+ */
+export function withEgressHandler (handler) {
+  return async (req, env, ctx) => {
+    if (env.FF_EGRESS_TRACKER_ENABLED !== 'true') {
+      return handler(req, env, ctx)
+    }
+
+    let response
+    try {
+      response = await handler(req, env, ctx)
+    } catch (error) {
+      console.error('Error in egress tracker handler:', error)
+      throw error
+    }
+
+    if (!response.ok || !response.body) {
+      return response
+    }
+
+    const { dataCid } = ctx
+    const accounting = ctx.ACCOUNTING_SERVICE ?? Accounting.create({
+      serviceURL: env.ACCOUNTING_SERVICE_URL
+    })
+
+    const { readable, writable } = createEgressPassThroughStream(ctx, accounting, dataCid)
+
+    try {
+      ctx.waitUntil(response.body.pipeTo(writable))
+    } catch (error) {
+      console.error('Error in egress tracker handler:', error)
+      // Original response in case of an error to avoid breaking the chain and serve the content
+      return response
+    }
+
+    return new Response(readable, {
+      status: response.status,
+      statusText: response.statusText,
+      headers: response.headers
+    })
+  }
+}
+
+/**
+ * Creates a TransformStream to count bytes served to the client.
+ * It records egress when the stream is finalized without an error.
+ *
+ * @param {import('@web3-storage/gateway-lib/middleware').Context} ctx - The context object.
+ * @param {AccountingService} accounting - The accounting service instance to record egress.
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} dataCid - The CID of the served content.
+ * @returns {TransformStream} - The created TransformStream.
```

Comment:
**suggestion:** Let's fully type this:

```ts
/**
 * @template {Uint8Array} T
 * @returns {TransformStream<T, T>} - The created TransformStream.
 */
```

That *should* also mean that you get to drop the explicit `@param` annotations inside the function, as the types should flow from there.